elements/lisk-tree/src/merkle_tree/calculate.ts

```diff
@@ -0,0 +1,47 @@
+/*
+ * Copyright © 2021 Lisk Foundation
+ *
+ * See the LICENSE file at the top-level directory of this distribution
+ * for licensing information.
+ *
+ * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
+ * no part of this software, including this file, may be copied, modified,
+ * propagated, or distributed except according to the terms contained in the
+ * LICENSE file.
+ *
+ * Removal or modification of this copyright notice is prohibited.
+ */
+
+import { hash } from '../../../lisk-cryptography/dist-node';
+import { LEAF_PREFIX } from './constants';
+import { calculatePathNodes } from './utils';
+
+export const calculateRootFromUpdateData = (
+	updateData: Buffer[],
+	proof: {
+		siblingHashes: Buffer[];
+		indexes: number[];
+		size: number;
+	},
+): Buffer => {
+	const { indexes, size, siblingHashes } = proof;
+
+	if (updateData.length !== indexes.length) {
+		throw new Error("Amount of update data doesn't match amount of indexes");
+	}
+
+	const updateHashes = [];
+
+	for (const data of updateData) {
+		const leafValueWithoutNodeIndex = Buffer.concat(
+			[LEAF_PREFIX, data],
+			LEAF_PREFIX.length + data.length,
+		);
+		const leafHash = hash(leafValueWithoutNodeIndex);
+		updateHashes.push(leafHash);
+	}
+
+	const calculatedTree = calculatePathNodes(updateHashes, size, indexes, siblingHashes);
+	const calculatedRoot = calculatedTree.get(2); // 2 is the index for root "10"
```

Comment:
In my understanding we now use indexes that are always encoded (with the prepended 1). Should we clarify this at some point? maybe at the beginning of the file in a comment or even better by choosing a different variable name for indexes? (maybe encodedIndexes?)

Reply:
Ill add a comment. For variable, we use
- index == encoded index
- nodeIndex, layerIndex => decoded index